sharktank/sharktank/ops/sharded_impls.py

```diff
@@ -30,6 +30,88 @@
+            name=tensor.name,
+            shape=tensor.shape,
+            devices=new_devices,
+            devices_pinned=tensor.devices_pinned
+        )
+    elif isinstance(tensor, ReplicatedTensor):
+        return ReplicatedTensor(
+            ts=new_shards,
+            name=tensor.name,
+            devices=new_devices,
+            devices_pinned=tensor.devices
+        )
+    elif isinstance(tensor, UnreducedTensor):
+        return UnreducedTensor(
+            ts=new_shards,
+            name=tensor.name,
+            shape=tensor.shape,
+            devices=new_devices
+        )
+    else:
+        raise NotImplementedError(f"copy_with_new_shards not implemented for {type(tensor)}")
+
+def transfer_if_needed(*tensors: Tuple[ShardedTensor, ...]) -> Tuple[ShardedTensor, ...]:
+    if len(tensors) <= 1:
+        return
+    assert all(isinstance(tensor, ShardedTensor) for tensor in tensors)
+    
+    # Check if all tensors are on the same devices.
+    devices_0 = tensors[0].devices
+    for tensor in tensors[1:]:
+        if not all(devices_0[j] == tensor.devices[j] for j in range(len(devices_0))):
+            break
+    else:
+        return tensors  # All tensors already on same device
+
+    i_pinned = tuple(i for i, t in enumerate(tensors) if t.devices_pinned)
+    if len(i_pinned) == 0:
+        raise ValueError("No tensors are pinned but some are on different devices, don't know where to transfer.")
+    
+    d_pinned = tensors[i_pinned[0]].devices
+    for i in i_pinned[1:]:
+        if not all(d_pinned[j] == tensors[i].devices[j] for j in range(len(d_pinned))):
+            raise ValueError("All pinned tensors must be on the same devices.")
+
+    # Move all non-pinned tensors to the same devices as the pinned ones.
+    new_tensors = []
+    for i, tensor in enumerate(tensors):
+        if tensor.devices_pinned:
+            new_tensors.append(tensor)
+        else:
+            shards = tuple(
+                (
+                    transfer_to_logical_device(shard, d_pinned[j])
+                    if d_pinned[j] != tensor.devices[j]
+                    else barrier_on_logical_device(shard, d_pinned[j])
+                )
+                for j, shard in enumerate(tensor.shards)
+            )
+            new_tensors.append(copy_w_new_shards_and_devices(tensor, shards, d_pinned))
+
+    return tuple(new_tensors)
+
+
+def override_w_tranfer(operation, *override_args):
+    def decorator(f):
+        @operation.override(*override_args)
+        def wrapper(*args: Tuple, **kwargs: Dict[str, Any]):
+            args = list(args)
+            t_i = [i for i, arg in enumerate(args) if isinstance(arg, ShardedTensor)]
+            if len(t_i) > 0:
+                t_vals = transfer_if_needed(*(args[i] for i in t_i))
+                for i, t in zip(t_i, t_vals):
+                    args[i] = t
+            return f(*args, **kwargs)  # TODO: Should I convert func_args back to tuple after modifying it?
```

Comment:
We may also want to add a special case that if the input is pinned and its an unary tensor operator we would also want to pin. Its a bit special but almost guaranteed.

Reply:
Are you thinking?:
```Python
res = f(*args, **kwargs)
if (
    len(args) == 1 
    and len(t_i) == 1
    and isinstance(res, ShardedTensor)
):
    res.pinned = True
return res
```

Reply:
Yest mostly. It should also check whether the input tensor was pinned. The idea is that if we do reshape/transpose on a pinned tensor its almost guaranteed we want the result pinned. We won't want it to pin a value that wasn't pinned though.

Reply:
Whoops messed up the logic there, yes on the same page about it.